stream.js

```diff
@@ -1,134 +1,155 @@
-//        else if (isDuplex(type))
-//          pull(_stream, err ? abortDuplex(err) : value, _stream)
+        //        if(isSource(type))
+        //          _stream(err ? pull.error(err) : value)
+        //        else if (isSink(type))
+        //          (err ? abortSink(err) : value)(_stream)
+        //        else if (isDuplex(type))
+        //          pull(_stream, err ? abortDuplex(err) : value, _stream)
       }
     },
 
     close: function (err) {
-        ps = null // deallocate
-        ws.ended = true
-        if(ws.closed) return
-        ws.closed = true
-        if(onClose) {
-          var close = onClose; onClose = null; close(err)
-        }
+      ps = null // deallocate
+      ws.ended = true
+      if (ws.closed) return
+      ws.closed = true
+      if (onClose) {
+        const close = onClose
+        onClose = null
+        close(err)
       }
+    }
   })
 
-  var ws = goodbye(pullWeird(ps, function () {
-    //this error will be handled in PacketStream.close
+  let ws = goodbye(pullWeird(ps, function () {
+    // this error will be handled in PacketStream.close
   }))
 
   ws = codec ? codec(ws) : ws
 
   ws.remoteCall = function (type, name, args, cb) {
-    if(name === 'emit') return ps.message(args)
+    if (name === 'emit') return ps.message(args)
 
-    if(!(isRequest(type) || isStream(type)))
+    if (!(isRequest(type) || isStream(type))) {
       throw new Error('unsupported type:' + JSON.stringify(type))
+    }
 
-    if(isRequest(type))
-      return ps.request({name: name, args: args}, cb)
+    if (isRequest(type)) {
+      return ps.request({ name: name, args: args }, cb)
+    }
 
-    var ws = ps.stream(), s = pullWeird[type](ws, cb)
-    ws.write({name: name, args: args, type: type})
+    const ws = ps.stream()
+    const s = pullWeird[type](ws, cb)
+    ws.write({ name: name, args: args, type: type })
     return s
   }
 
-
-  //hack to work around ordering in setting ps.ended.
-  //Question: if an object has subobjects, which
-  //all have close events, should the subobjects fire close
-  //before the parent? or should parents close after?
-  //should there be a preclose event on the parent
-  //that fires when it's about to close all the children?
+  // hack to work around ordering in setting ps.ended.
+  // Question: if an object has subobjects, which
+  // all have close events, should the subobjects fire close
+  // before the parent? or should parents close after?
+  // should there be a preclose event on the parent
+  // that fires when it's about to close all the children?
   ws.isOpen = function () {
     return !ps.ended
   }
 
   ws.close = function (err, cb) {
-    if(isFunction(err))
-      cb = err, err = false
-    if(!ps) return (cb && cb())
-    if(err) return ps.destroy(err), (cb && cb())
+    if (isFunction(err)) {
+      cb = err
+      err = false
+    }
+    if (!ps) {
+      if (cb) cb()
+      return
+    }
+    if (err) {
+      if (cb) cb()
+      ps.destroy(err)
```

Comment:
this is a great example of why this cleanup is great. The way I read the old code was that destroy is called first and afterwards cb. Here they are changed? Is that correct?

Reply:
Hmm, sorry! I'll swap them back. I thought that the author's intent was to `return ps.destroy(err)` and then call `if (cb) cb()` **after** the return, but I realized that that's not possible, and also that `ps.destroy(err)` returns undefined, so really, this block should just be

- `ps.destroy(err)`
- `if (cb) cb()`
- return

PS: running code after `return` is possible, just not with the patterns here (neither before nor after the git diff). It's possible with a `finally` block.

Reply:
Fixed this one